server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -90,6 +91,7 @@
     private volatile Map<String, Pipeline> pipelines = new HashMap<>();
     private final ThreadPool threadPool;
     private final IngestMetric totalMetrics = new IngestMetric();
+    private final List<Consumer<ClusterState>> ingestClusterStateListeners = new CopyOnWriteArrayList<>();
```

Comment:
should this be a Set to avoid any expectations of the listeners getting applied in a specific order ? (since they are applied in the order in which they are added, and that could be non-deterministic)

Reply:
My (weak) preference is to leave as is, as we are using this pattern in other places of the listener infrastructure as well.